BedrockServer.cpp

```diff
@@ -57,217 +68,346 @@ void BedrockServer_WorkerThread_ProcessDirectMessages(BedrockNode& node, Bedrock
+            SAUTOPREFIX(command->request["requestID"]);
+            escalatedCommand = true;
+            command = node.reopenCommand(command);
+        } else {
+            // Otherwise, let's see if we can get a new request.
+            SData request = data.queuedRequests.pop();
 
-            int maxS = node.preSelect(fdm);
-            maxS = max(queuedEscalatedRequests.preSelect(fdm), maxS);
-            maxS = max(directMessages.preSelect(fdm), maxS);
-            const uint64_t now = STimeNow();
-            data->server->pollTimer.start();
-            S_poll(fdm, max(nextActivity, now) - now);
-            data->server->pollTimer.stop();
-            nextActivity = STimeNow() + STIME_US_PER_S; // 1s max period
-
-            // Handle any HTTPS requests from our plugins.
-            for (list<SHTTPSManager*>& managerList : httpsManagers) { 
-                for (SHTTPSManager* manager : managerList) {
-                    manager->postSelect(fdm, nextActivity);
-                }
+            // If we didn't get anything here, there's no work to do. Go back to waiting.
+            if (request.empty()) {
+                continue;
             }
 
-            node.postSelect(fdm, nextActivity);
-            queuedEscalatedRequests.postSelect(fdm);
-            directMessages.postSelect(fdm);
-
-            // Process any direct messages from the main thread to us
-            BedrockServer_WorkerThread_ProcessDirectMessages(node, directMessages);
+            // This *needs* to be after the empty check, because otherwise we'll have added an empty 'requestID' to the
+            // request, making it non-empty.
+            SAUTOPREFIX(request["requestID"]);
 
-            // Check for available work.
-            while (true) {
-                // Try to get some work
-                const SData& request = queuedEscalatedRequests.pop();
-                if (request.empty())
-                    break;
+            // If the command is scheduled for the future, we'll forward it to the sync thread, as only the sync thread
+            // keeps a long-running queue instead of operating on one command at a time.
+            // Also, if the command is a status command, we'll forward it to the sync thread, because status commands
+            // are special and require access to information that only the sync thread knows.
+            command = node.createCommand(request);
+            if(command->creationTimestamp > STimeNow() ||
+               find(BedrockPlugin_Status::statusCommandNames.begin(), BedrockPlugin_Status::statusCommandNames.end(),
+                    request.methodLine) != BedrockPlugin_Status::statusCommandNames.end()) {
+                SINFO("Forwarding command " << command->id << " to sync thread.");
+                command->response.clear(); // TODO: These should be clear on creation.
+                data.peekedCommands.push(command);
```

Comment:
No, the sync thread will always `peek` it anyway, it's just a bad name for the queue (the name was chosen before I addressed this condition).